### PR TITLE
backtest : fix ftx backtest for #355

### PR DESCRIPTION
**Pull request by tony1223**

### CODE REVIEW

File: pkg/bbgo/config.go [hunk 1]
```diff
@@ -476,6 +476,9 @@ func loadExchangeStrategies(config *Config, stash Stash) (err error) {
 					Mounts:   mounts,
 					Strategy: st,
 				})
+			} else if id != "on" && id != "off" {
+				//Show error when we didn't find the Strategy
+				logrus.Errorf(" Strategy %s in config not found.", id)
```
Comment (c9s) [inline pkg/bbgo/config.go:481]: error messages are in lower cases~

File: pkg/cmd/trades.go [hunk 2]
```diff
@@ -101,15 +101,17 @@ var tradesCmd = &cobra.Command{
...
 				trade.Symbol,
 				trade.Side,
 				util.FormatFloat(trade.Quantity, 4),
 				util.FormatFloat(trade.Price, 3),
 				trade.OrderID,
 				trade.Time.Time().Format(time.StampMilli),
-				trade.QuoteQuantity)
+				trade.QuoteQuantity,
+				trade.FeeCurrency,
+				trade.Fee)
```
Comment (c9s) [inline pkg/cmd/trades.go:114]: use trade.Fee then trade.FeeCurrency, so it show display something like 0.001 BNB ? looks better?

File: pkg/exchange/batch/batch.go [hunk 3]
```diff
@@ -74,18 +76,21 @@ type KLineBatchQuery struct {
 	types.Exchange
 }
 
-func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval types.Interval, startTime, endTime time.Time) (c chan types.KLine, errC chan error) {
-	c = make(chan types.KLine, 1000)
+func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval types.Interval, startTime, endTime time.Time) (c chan []types.KLine, errC chan error) {
+	c = make(chan []types.KLine, 1000)
```
Comment (c9s) [inline pkg/exchange/batch/batch.go:80]: why changing it to `[]types.KLine`?

File: pkg/exchange/ftx/exchange.go [hunk 4]
```diff
@@ -179,7 +179,70 @@ func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap,
 	return balances, nil
 }
 
+func (e *Exchange) SupportedInterval() map[types.Interval]int {
+	return map[types.Interval]int{
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:183]: pull out to a global ? var supportedIntervals = map[types.Interval]int{....} something like this, and you can reuse them.

File: pkg/exchange/batch/batch.go (hunk 3 above)
Comment (tony1223) [reply pkg/exchange/batch/batch.go:80]: for batch insert purpose.

File: pkg/bbgo/config.go (hunk 1 above)
Comment (tony1223) [reply pkg/bbgo/config.go:481]: ok

File: pkg/exchange/ftx/exchange.go (hunk 4 above)
Comment (tony1223) [reply pkg/exchange/ftx/exchange.go:183]: ok

File: pkg/service/backtest.go [hunk 5]
```diff
@@ -138,25 +158,40 @@ func (s *BacktestService) QueryKLinesBackward(exchange types.ExchangeName, symbo
 }
 
 func (s *BacktestService) QueryKLinesCh(since, until time.Time, exchange types.Exchange, symbols []string, intervals []types.Interval) (chan types.KLine, chan error) {
-	sql := "SELECT * FROM `binance_klines` WHERE `end_time` BETWEEN :since AND :until AND `symbol` IN (:symbols) AND `interval` IN (:intervals) ORDER BY end_time ASC"
-	sql = strings.ReplaceAll(sql, "binance_klines", exchange.Name().String()+"_klines")
+
+	if len(symbols) == 0 {
+
+		log.Error("backtest: symbols is empty when querying kline, plesae check your strategy setting. ")
```
Comment (c9s) [inline pkg/service/backtest.go:164]: errC is sending the same error message, so this one can be removed?

File: pkg/service/backtest.go [hunk 6]
```diff
@@ -211,14 +246,47 @@ func (s *BacktestService) scanRows(rows *sqlx.Rows) (klines []types.KLine, err e
 	return klines, rows.Err()
 }
 
+func (s *BacktestService) _targetKlineTable(exchangeName types.ExchangeName) string {
+	if exchangeName == types.ExchangeBinance {
```
Comment (c9s) [inline pkg/service/backtest.go:250]: use switch case here?

File: pkg/types/trade.go [hunk 7]
```diff
@@ -87,13 +87,15 @@ func (trade Trade) String() string {
...
+	return fmt.Sprintf("Trade %s %s %s %f @ %f, amount %f , fee %s %f ",
 		trade.Exchange.String(),
 		trade.Symbol,
 		trade.Side,
 		trade.Quantity,
 		trade.Price,
-		trade.QuoteQuantity)
+		trade.QuoteQuantity,
+		trade.FeeCurrency,
+		trade.Fee)
```
Comment (c9s) [inline pkg/types/trade.go:98]: using `trade.Fee trade.Currency` might be better

**Comment by c9s** [comment]
test failed

File: pkg/service/backtest.go (hunk 5 above)
Comment (tony1223) [reply pkg/service/backtest.go:164]: agree

File: pkg/types/trade.go (hunk 7 above)
Comment (tony1223) [reply pkg/types/trade.go:98]: ok

File: pkg/service/backtest.go (hunk 6 above)
Comment (tony1223) [reply pkg/service/backtest.go:250]: ok

File: pkg/exchange/batch/batch.go [hunk 8]
```diff
@@ -106,8 +112,23 @@ func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval type
 					return
 				}
 
-				c <- kline
-				startTime = kline.EndTime.Add(time.Millisecond)
+				workingKLines = append(workingKLines, kline)
+
+				if len(workingKLines) == 200 {
```
Comment (c9s) [inline pkg/exchange/batch/batch.go:117]: pull 200 out to a constant?

File: pkg/exchange/batch/batch.go [hunk 9]
```diff
@@ -96,6 +101,7 @@ func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval type
 				return
 			}
 
+			var workingKLines []types.KLine
```
Comment (c9s) [inline pkg/exchange/batch/batch.go:104]:
```
var workingKLines = make([]types.KLine, 0, NumofBatchKLine)
```
pre-allocate it?

File: pkg/exchange/batch/batch.go [hunk 10]
```diff
@@ -106,8 +112,23 @@ func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval type
 					return
 				}
 
-				c <- kline
-				startTime = kline.EndTime.Add(time.Millisecond)
+				workingKLines = append(workingKLines, kline)
+
+				if len(workingKLines) == 200 {
+					c <- workingKLines
+					workingKLines = nil
```
Comment (c9s) [inline pkg/exchange/batch/batch.go:119]: `workingKLines = workingKLines[:0]`

this preserve the underlying memory but reset the slice.

File: pkg/exchange/batch/batch.go [hunk 11]
```diff
@@ -106,8 +112,23 @@ func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval type
...
+				}
+
+				//The issue is in FTX, prev endtime = next start time , so if add 1 ms , it would query forever.
+				startTime = kline.EndTime // .Add(time.Millisecond)
+				tryQueryKlineTimes = 0
+			}
+
+			c <- workingKLines
```
Comment (c9s) [inline pkg/exchange/batch/batch.go:127]: rename workingKLines to `batchKLines` might make it easier to understand its purpose.

File: pkg/exchange/ftx/exchange.go [hunk 12]
```diff
@@ -179,7 +179,72 @@ func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap,
...
+	}
+	if options.EndTime != nil {
+		until = *options.EndTime
+	} else {
+		until = time.Now()
+	}
+
+	current = until
+
+	for true {
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:213]: can be just `for {`

File: pkg/exchange/ftx/exchange.go [hunk 13]
```diff
@@ -179,7 +179,72 @@ func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap,
 	return balances, nil
 }
 
+var supportedIntervals = map[types.Interval]int{
+	types.Interval1m:  1,
+	types.Interval5m:  5,
+	types.Interval15m: 15,
+	types.Interval1h:  60,
+	types.Interval1d:  60 * 24,
+	types.Interval3d:  60 * 24 * 3,
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:188]: 30m and 6h, 12h are not supported?

File: pkg/exchange/ftx/exchange.go [hunk 14]
```diff
@@ -203,25 +273,56 @@ func (e *Exchange) QueryKLines(ctx context.Context, symbol string, interval type
...
+	345600:  {},
+	432000:  {},
+	518400:  {},
+	604800:  {},
+	691200:  {},
+	777600:  {},
+	864000:  {},
+	950400:  {},
+	1036800: {},
+	1123200: {},
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:308]: maybe this can be merged into `supportedIntervals`? (I am not sure)

File: pkg/service/backtest.go [hunk 15]
```diff
@@ -35,12 +36,16 @@ func (s *BacktestService) SyncKLineByInterval(ctx context.Context, exchange type
 
 	// should use channel here
 	klineC, errC := batch.Query(ctx, symbol, interval, startTime, endTime)
+
 	// var previousKLine types.KLine
+	count := 0
 	for k := range klineC {
```
Comment (c9s) [inline pkg/service/backtest.go:42]: rename k -> ks ? since it's plural?

File: pkg/exchange/ftx/exchange.go [hunk 16]
```diff
@@ -179,7 +179,72 @@ func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap,
...
+
+	current = until
+
+	for true {
+
+		endTime := current.Add(interval.Duration())
+		options.EndTime = &endTime
+		lines, err := e._queryKLines(ctx, symbol, interval, options)
+
+		if len(lines) == 0 {
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:219]: if len(lines) == 0 , but err is not nil ?

pre-check the error before this if block?

File: pkg/cmd/trades.go (hunk 2 above)
Comment (tony1223) [reply pkg/cmd/trades.go]: ok

File: pkg/exchange/ftx/exchange.go (hunk 12 above)
Comment (tony1223) [reply pkg/exchange/ftx/exchange.go:213]: ok

File: pkg/exchange/batch/batch.go (hunk 11 above)
Comment (tony1223) [reply pkg/exchange/batch/batch.go:127]: ok

File: pkg/exchange/batch/batch.go (hunk 10 above)
Comment (tony1223) [reply pkg/exchange/batch/batch.go:119]: ok

File: pkg/exchange/batch/batch.go (hunk 9 above)
Comment (tony1223) [reply pkg/exchange/batch/batch.go:104]: ok

File: pkg/exchange/batch/batch.go (hunk 8 above)
Comment (tony1223) [reply pkg/exchange/batch/batch.go:117]: ok

File: pkg/exchange/ftx/exchange.go (hunk 14 above)
Comment (tony1223) [reply pkg/exchange/ftx/exchange.go:308]: yes, we  could do that to limit the options to only supprotedIntervals items.  the issue is the config range different than types.interval .

File: pkg/service/backtest.go (hunk 15 above)
Comment (tony1223) [reply pkg/service/backtest.go:42]: i renamed it to "klines"

File: pkg/exchange/ftx/exchange.go (hunk 16 above)
Comment (tony1223) [reply pkg/exchange/ftx/exchange.go:219]: ok

File: pkg/exchange/ftx/exchange.go (hunk 13 above)
Comment (tony1223) [reply pkg/exchange/ftx/exchange.go:190]: yes ,

according to the document

https://docs.ftx.com/#get-historical-index
resolution field
//window length in seconds. options: 15, 60, 300, 900, 3600, 14400, 86400, or any multiple of 86400 up to 30*86400

which means it only support the following interval
0.25m,1m,5m,15m,1h,4h,1d,2d...30d

File: pkg/exchange/batch/batch.go [hunk 17]
```diff
@@ -95,7 +100,9 @@ func (e KLineBatchQuery) Query(ctx context.Context, symbol string, interval type
 			if len(kLines) == 0 {
 				return
 			}
+			const BatchSize = 200
 
+			var batchKLines = make([]types.KLine, 0, (len(kLines)/BatchSize)+1)
```
Comment (c9s) [inline pkg/exchange/batch/batch.go:105]: since you will reset batchKLines when len == BatchSize, so the capacity should be BatchSize ?

**Review by c9s — APPROVED** [review]

File: pkg/bbgo/config.go [hunk 18]
```diff
@@ -476,6 +476,9 @@ func loadExchangeStrategies(config *Config, stash Stash) (err error) {
 					Mounts:   mounts,
 					Strategy: st,
 				})
+			} else if id != "on" && id != "off" {
+				//Show error when we didn't find the Strategy
+				logrus.Errorf("strategy %s in config not found.", id)
```
Comment (c9s) [inline pkg/bbgo/config.go:481]: you can return fmt.Errorf() as an error

**Review by c9s — APPROVED** [review]

**Review by c9s — APPROVED** [review]

**Pull request merged**